operator/duties/proposer.go

```diff
@@ -0,0 +1,177 @@
+//
+// On Ticker event:
+//  1. Execute duties.
+//  2. If necessary, fetch duties for the current epoch.
+func (h *ProposerHandler) HandleDuties(ctx context.Context, logger *zap.Logger) {
+	logger = logger.With(zap.String("handler", h.Name()))
+	logger.Info("starting duty handler")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case slot := <-h.ticker:
+			currentEpoch := h.network.Beacon.EstimatedEpochAtSlot(slot)
+			buildStr := fmt.Sprintf("e%v-s%v-#%v", currentEpoch, slot, slot%32+1)
+			logger.Debug("🛠 ticker event", zap.String("epoch_slot_seq", buildStr))
+
+			if h.fetchFirst {
+				h.fetchFirst = false
+				h.processFetching(ctx, logger, currentEpoch, slot)
+				h.processExecution(logger, currentEpoch, slot)
+			} else {
+				h.processExecution(logger, currentEpoch, slot)
+				if h.indicesChanged {
+					h.duties.Reset(currentEpoch)
+					h.indicesChanged = false
+					h.processFetching(ctx, logger, currentEpoch, slot)
+				}
+			}
+
+			// last slot of epoch
+			if uint64(slot)%h.network.Beacon.SlotsPerEpoch() == h.network.Beacon.SlotsPerEpoch()-1 {
+				h.duties.Reset(currentEpoch)
+				h.fetchFirst = true
+			}
+
+		case reorgEvent := <-h.reorg:
+			currentEpoch := h.network.Beacon.EstimatedEpochAtSlot(reorgEvent.Slot)
+			buildStr := fmt.Sprintf("e%v-s%v-#%v", currentEpoch, reorgEvent.Slot, reorgEvent.Slot%32+1)
+			logger.Info("🔀 reorg event received", zap.String("epoch_slot_seq", buildStr), zap.Any("event", reorgEvent))
+
+			// reset current epoch duties
+			if reorgEvent.Current {
+				h.duties.Reset(currentEpoch)
+				h.fetchFirst = true
+			}
+
+		case <-h.indicesChange:
+			slot := h.network.Beacon.EstimatedCurrentSlot()
+			currentEpoch := h.network.Beacon.EstimatedEpochAtSlot(slot)
+			buildStr := fmt.Sprintf("e%v-s%v-#%v", currentEpoch, slot, slot%32+1)
+			logger.Info("🔁 indices change received", zap.String("epoch_slot_seq", buildStr))
+
+			h.indicesChanged = true
+		}
+	}
+}
+
+func (h *ProposerHandler) processFetching(ctx context.Context, logger *zap.Logger, epoch phase0.Epoch, slot phase0.Slot) {
+	ctx, cancel := context.WithDeadline(ctx, h.network.Beacon.GetSlotStartTime(slot+1).Add(100*time.Millisecond))
+	defer cancel()
+
+	if err := h.fetchDuties(ctx, logger, epoch); err != nil {
+		logger.Error("failed to fetch duties for current epoch", zap.Error(err))
+		return
+	}
+}
+
+func (h *ProposerHandler) processExecution(logger *zap.Logger, epoch phase0.Epoch, slot phase0.Slot) {
+	// range over duties and execute
+	if slotMap, ok := h.duties.m[epoch]; ok {
+		if duties, ok := slotMap[slot]; ok {
+			toExecute := make([]*spectypes.Duty, 0, len(duties))
+			for _, d := range duties {
+				if h.shouldExecute(logger, d) {
+					toExecute = append(toExecute, h.toSpecDuty(d, spectypes.BNRoleProposer))
+				}
+			}
+			h.executeDuties(logger, toExecute)
+		}
+	}
+}
+
+func (h *ProposerHandler) fetchDuties(ctx context.Context, logger *zap.Logger, epoch phase0.Epoch) error {
```

Comment:
I think fetchDuties should only fetch duties, return them and the calling function should decide what to do with it (ex: add to the map, log etc..). otherwise this function name should be different.

Reply:
we want to have one place to have it all (the fetch, the log the subscribe and so on)
we can rename it to fetchAndProcessDuties